numcodecs/msgpacks.py

```diff
@@ -52,7 +52,10 @@ def __init__(self, use_single_float=False, use_bin_type=True, raw=False):
         self.raw = raw
 
     def encode(self, buf):
-        buf = np.asarray(buf)
+        try:
+            buf = np.asarray(buf)
+        except ValueError:
+            buf = np.asarray(buf, dtype=object)
```

Comment:
```suggestion
        buf = np.asarray(buf, dtype=object)
```

because in previous versions of numpy, you ended up with object dtype anyway, so nothing is lost.

Reply:
That was my first try, but it is not true for non-ragged arrays. See also https://github.com/zarr-developers/numcodecs/pull/416#issuecomment-1380729833

Tests fails when you do this:

```
[   55s] =================================== FAILURES ===================================
[   55s] ______________________________ test_encode_decode ______________________________
[   55s] 
[   55s]     def test_encode_decode():
[   55s]         for arr in arrays:
[   55s] >           check_encode_decode_array(arr, MsgPack())
[   55s] 
[   55s] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_msgpacks.py:39: 
[   55s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[   55s] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/common.py:209: in check_encode_decode_array
[   55s]     assert_array_items_equal(arr, dec)
[   55s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[   55s] 
[   55s] res = array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
[   55s]         13,  14,  15,  16,  17,  18,  19,  20,...1, 982, 983, 984, 985, 986, 987,
[   55s]        988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999],
[   55s]       dtype=int32)
[   55s] arr = array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
[   55s]        19, 20, 21, 22, 23, 24, 25, 26, 27, 28..., 982, 983, 984, 985, 986,
[   55s]        987, 988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999],
[   55s]       dtype=object)
[   55s] 
[   55s]     def assert_array_items_equal(res, arr):
[   55s]     
[   55s]         assert isinstance(res, np.ndarray)
[   55s]         res = res.reshape(-1, order='A')
[   55s]         arr = arr.reshape(-1, order='A')
[   55s]         assert res.shape == arr.shape
[   55s] >       assert res.dtype == arr.dtype
[   55s] E       AssertionError: assert dtype('int32') == dtype('O')
[   55s] E        +  where dtype('int32') = array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,\n        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,\n        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,\n        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,\n        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,\n        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,\n        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,\n        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,\n       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,\n       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,\n       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,\n       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,\n       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,\n       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181,\n       182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194,\n       195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207,\n       208, 209, 210, 211, 212, 213, 214, 215,...6, 787, 788, 789, 790, 791, 792,\n       793, 794, 795, 796, 797, 798, 799, 800, 801, 802, 803, 804, 805,\n       806, 807, 808, 809, 810, 811, 812, 813, 814, 815, 816, 817, 818,\n       819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830, 831,\n       832, 833, 834, 835, 836, 837, 838, 839, 840, 841, 842, 843, 844,\n       845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857,\n       858, 859, 860, 861, 862, 863, 864, 865, 866, 867, 868, 869, 870,\n       871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882, 883,\n       884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895, 896,\n       897, 898, 899, 900, 901, 902, 903, 904, 905, 906, 907, 908, 909,\n       910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922,\n       923, 924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934, 935,\n       936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948,\n       949, 950, 951, 952, 953, 954, 955, 956, 957, 958, 959, 960, 961,\n       962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973, 974,\n       975, 976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986, 987,\n       988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999],\n      dtype=int32).dtype
[   55s] E        +  and   dtype('O') = array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,\n       19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,\n       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,\n       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,\n       70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86,\n       87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102,\n       103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115,\n       116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128,\n       129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141,\n       142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154,\n       155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167,\n       168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180,\n       181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193,\n       194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206,\n       207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219,\n       220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232,\n       233, 234, 235, 236, 237,..., 787, 788, 789, 790, 791,\n       792, 793, 794, 795, 796, 797, 798, 799, 800, 801, 802, 803, 804,\n       805, 806, 807, 808, 809, 810, 811, 812, 813, 814, 815, 816, 817,\n       818, 819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830,\n       831, 832, 833, 834, 835, 836, 837, 838, 839, 840, 841, 842, 843,\n       844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856,\n       857, 858, 859, 860, 861, 862, 863, 864, 865, 866, 867, 868, 869,\n       870, 871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882,\n       883, 884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895,\n       896, 897, 898, 899, 900, 901, 902, 903, 904, 905, 906, 907, 908,\n       909, 910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921,\n       922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934,\n       935, 936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947,\n       948, 949, 950, 951, 952, 953, 954, 955, 956, 957, 958, 959, 960,\n       961, 962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973,\n       974, 975, 976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986,\n       987, 988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999],\n      dtype=object).dtype
[   55s] 
[   55s] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/common.py:189: AssertionError
[   55s] ___________________ test_encode_decode_shape_dtype_preserved ___________________
[   55s] 
[   55s]     def test_encode_decode_shape_dtype_preserved():
[   55s]         codec = MsgPack()
[   55s]         for arr in arrays:
[   55s]             actual = codec.decode(codec.encode(arr))
[   55s]             assert arr.shape == actual.shape
[   55s] >           assert arr.dtype == actual.dtype
[   55s] E           AssertionError: assert dtype('int32') == dtype('O')
[   55s] E            +  where dtype('int32') = array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,\n        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,\n        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,\n        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,\n        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,\n        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,\n        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,\n        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,\n       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,\n       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,\n       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,\n       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,\n       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,\n       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181,\n       182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194,\n       195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207,\n       208, 209, 210, 211, 212, 213, 214, 215,...6, 787, 788, 789, 790, 791, 792,\n       793, 794, 795, 796, 797, 798, 799, 800, 801, 802, 803, 804, 805,\n       806, 807, 808, 809, 810, 811, 812, 813, 814, 815, 816, 817, 818,\n       819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830, 831,\n       832, 833, 834, 835, 836, 837, 838, 839, 840, 841, 842, 843, 844,\n       845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857,\n       858, 859, 860, 861, 862, 863, 864, 865, 866, 867, 868, 869, 870,\n       871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882, 883,\n       884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895, 896,\n       897, 898, 899, 900, 901, 902, 903, 904, 905, 906, 907, 908, 909,\n       910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922,\n       923, 924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934, 935,\n       936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948,\n       949, 950, 951, 952, 953, 954, 955, 956, 957, 958, 959, 960, 961,\n       962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973, 974,\n       975, 976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986, 987,\n       988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999],\n      dtype=int32).dtype
[   55s] E            +  and   dtype('O') = array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,\n       19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,\n       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,\n       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,\n       70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86,\n       87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102,\n       103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115,\n       116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128,\n       129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141,\n       142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154,\n       155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167,\n       168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180,\n       181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193,\n       194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206,\n       207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219,\n       220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232,\n       233, 234, 235, 236, 237,..., 787, 788, 789, 790, 791,\n       792, 793, 794, 795, 796, 797, 798, 799, 800, 801, 802, 803, 804,\n       805, 806, 807, 808, 809, 810, 811, 812, 813, 814, 815, 816, 817,\n       818, 819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830,\n       831, 832, 833, 834, 835, 836, 837, 838, 839, 840, 841, 842, 843,\n       844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856,\n       857, 858, 859, 860, 861, 862, 863, 864, 865, 866, 867, 868, 869,\n       870, 871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882,\n       883, 884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895,\n       896, 897, 898, 899, 900, 901, 902, 903, 904, 905, 906, 907, 908,\n       909, 910, 911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921,\n       922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934,\n       935, 936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947,\n       948, 949, 950, 951, 952, 953, 954, 955, 956, 957, 958, 959, 960,\n       961, 962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973,\n       974, 975, 976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986,\n       987, 988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999],\n      dtype=object).dtype
[   55s] 
[   55s] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_msgpacks.py:89: AssertionError
[   55s] =========================== short test summary info ============================
[   55s] SKIPPED [2] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_zfpy.py:11: ZFPY not available
[   55s] SKIPPED [28] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_blosc.py:66: codec has been removed
[   55s] SKIPPED [1] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/common.py:265: codec has been removed
[   55s] SKIPPED [1] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_ndarray_like.py:8: could not import 'cupy': No module named 'cupy'
[   55s] SKIPPED [1] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_ndarray_like.py:20: could not import 'cupy': No module named 'cupy'
[   55s] SKIPPED [1] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_ndarray_like.py:32: could not import 'cupy': No module named 'cupy'
[   55s] SKIPPED [1] ../../BUILDROOT/python-numcodecs-0.11.0-0.x86_64/usr/lib64/python3.8/site-packages/numcodecs/tests/test_ndarray_like.py:44: could not import 'cupy': No module named 'cupy'
[   55s] FAILED tests/test_msgpacks.py::test_encode_decode - AssertionError: assert dt...
[   55s] FAILED tests/test_msgpacks.py::test_encode_decode_shape_dtype_preserved - Ass...
[   55s] ============ 2 failed, 566 passed, 35 skipped, 38 xfailed in 13.17s ============
```

Reply:
right, OK - sorry.